File: src/protocol/metadata.rs
use std::collections::HashMap;

use crate::protocol::constants::*;
use crate::protocol::types::PacketType;
use crate::protocol::{Packet, PacketError};

/// A single metadata entry: key-value pair.
#[derive(Debug, Clone)]
pub struct MetadataEntry {
    pub key: u8,
    pub value: Vec<u8>,
}

/// Metadata block containing one or more entries.
#[derive(Debug, Clone, Default)]
pub struct Metadata {
    entries: Vec<MetadataEntry>,
}

/// Type alias for convenience building metadata.
pub type MetadataMap = HashMap<u8, Vec<u8>>;

impl Metadata {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn from_map(map: &MetadataMap) -> Self {
        let mut m = Self::new();
        for (&key, value) in map {
            m.set(key, value.clone());
        }
        m
    }

    pub fn set(&mut self, key: u8, value: Vec<u8>) {
        if let Some(entry) = self.entries.iter_mut().find(|e| e.key == key) {
            entry.value = value;
        } else {
            self.entries.push(MetadataEntry { key, value });
        }
    }

    pub fn set_string(&mut self, key: u8, value: &str) {
        self.set(key, value.as_bytes().to_vec());
    }

    pub fn set_u32(&mut self, key: u8, value: u32) {
        self.set(key, value.to_be_bytes().to_vec());
    }

    pub fn set_u16(&mut self, key: u8, value: u16) {
        self.set(key, value.to_be_bytes().to_vec());
    }

    pub fn get(&self, key: u8) -> Option<&[u8]> {
        self.entries
            .iter()
            .find(|e| e.key == key)
            .map(|e| e.value.as_slice())
    }

    pub fn get_string(&self, key: u8) -> Option<String> {
        self.get(key)
            .and_then(|v| std::str::from_utf8(v).ok())
            .map(String::from)
    }

    pub fn get_u32(&self, key: u8) -> Option<u32> {
        self.get(key).and_then(|v| {
            if v.len() >= 4 {
                let mut b = [0u8; 4];
                b.copy_from_slice(&v[..4]);
                Some(u32::from_be_bytes(b))
            } else {
                None
            }
        })
    }

    pub fn get_u16(&self, key: u8) -> Option<u16> {
        self.get(key).and_then(|v| {
            if v.len() >= 2 {
                let mut b = [0u8; 2];
                b.copy_from_slice(&v[..2]);
                Some(u16::from_be_bytes(b))
            } else {
                None
            }
        })
    }

    pub fn remove(&mut self, key: u8) -> bool {
        let len_before = self.entries.len();
        self.entries.retain(|e| e.key != key);
        self.entries.len() < len_before
    }

    pub fn contains(&self, key: u8) -> bool {
        self.entries.iter().any(|e| e.key == key)
    }

    pub fn entries(&self) -> &[MetadataEntry] {
        &self.entries
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn iter(&self) -> impl Iterator<Item = (&u8, &[u8])> {
        self.entries.iter().map(|e| (&e.key, e.value.as_slice()))
    }

    /// Encode all entries into a byte vector.
    /// Wire format per entry: [key: u8] [length: u16 BE] [value: N bytes]
    pub fn encode(&self) -> Vec<u8> {
        let total_size: usize = self
            .entries
            .iter()
            .map(|e| METADATA_ENTRY_HEADER_SIZE + e.value.len())
            .sum();
        let mut buf = Vec::with_capacity(total_size);
        for entry in &self.entries {
            buf.push(entry.key);
            buf.extend_from_slice(&(entry.value.len() as u16).to_be_bytes());
            buf.extend_from_slice(&entry.value);
        }
        buf
    }

    /// Decode entries from a byte slice.
    pub fn decode(data: &[u8]) -> Result<Self, PacketError> {
        let mut entries = Vec::new();
        let mut offset = 0;

        while offset < data.len() {
            if data.len() - offset < METADATA_ENTRY_HEADER_SIZE {
                return Err(PacketError::InvalidPayload("metadata entry truncated"));
            }

            let key = data[offset];
            offset += 1;

            let mut len_bytes = [0u8; 2];
            len_bytes.copy_from_slice(&data[offset..offset + 2]);
            let value_len = u16::from_be_bytes(len_bytes) as usize;
            offset += 2;

            if offset + value_len > data.len() {
                return Err(PacketError::InvalidPayload("metadata value truncated"));
            }

            let value = data[offset..offset + value_len].to_vec();
            offset += value_len;

            entries.push(MetadataEntry { key, value });
        }

        Ok(Self { entries })
    }

    /// Wrap into a Packet with PacketType::Metadata.
    pub fn to_packet(&self, seq: u32, ts: u32) -> Packet {
        Packet::new(PacketType::Metadata, seq, ts, self.encode())
    }

    /// Extract Metadata from a Packet.
    pub fn from_packet(packet: &Packet) -> Result<Self, PacketError> {
        if packet.header.packet_type != PacketType::Metadata {
            return Err(PacketError::InvalidPayload("not a metadata packet"));
        }
        Self::decode(&packet.payload)
    }
}

/// Convenience builder for common metadata fields.
pub struct MetadataBuilder {
    meta: Metadata,
}

impl MetadataBuilder {
    pub fn new() -> Self {
        Self {
            meta: Metadata::new(),
        }
    }

    pub fn ssrc(mut self, ssrc: u32) -> Self {
        self.meta.set_u32(META_SSRC, ssrc);
        self
    }

    pub fn track_title(mut self, title: &str) -> Self {
        self.meta.set_string(META_TRACK_TITLE, title);
        self
    }

    pub fn artist(mut self, artist: &str) -> Self {
        self.meta.set_string(META_ARTIST, artist);
        self
    }

    pub fn album(mut self, album: &str) -> Self {
        self.meta.set_string(META_ALBUM, album);
        self
    }

    pub fn genre(mut self, genre: &str) -> Self {
        self.meta.set_string(META_GENRE, genre);
        self
    }

    pub fn sample_rate(mut self, rate: u32) -> Self {
        self.meta.set_u32(META_SAMPLE_RATE, rate);
        self
    }

    pub fn channels(mut self, channels: u16) -> Self {
        self.meta.set_u16(META_CHANNELS, channels);
        self
    }

    pub fn codec_info(mut self, info: &str) -> Self {
        self.meta.set_string(META_CODEC_INFO, info);
        self
    }

    pub fn bitrate(mut self, bitrate: u32) -> Self {
        self.meta.set_u32(META_BITRATE, bitrate);
        self
    }

    pub fn duration_ms(mut self, duration: u64) -> Self {
        self.meta
            .set(key_u64(META_DURATION_MS), duration.to_be_bytes().to_vec());
        self
    }

    pub fn stream_title(mut self, title: &str) -> Self {
        self.meta.set_string(META_STREAM_TITLE, title);
        self
    }

    pub fn stream_url(mut self, url: &str) -> Self {
        self.meta.set_string(META_STREAM_URL, url);
        self
    }

    pub fn custom(mut self, key: u8, value: Vec<u8>) -> Self {
        if key < META_CUSTOM_BASE {
            panic!("custom metadata key must be >= 0x80");
        }
        self.meta.set(key, value);
        self
    }

    pub fn build(self) -> Metadata {
        self.meta
    }
}

fn key_u64(key: u8) -> u8 {
    key
}

impl Default for MetadataBuilder {
    fn default() -> Self {
        Self::new()
    }
}

File: ports/python/rushaudio/metadata.py
"""Metadata TLV encode/decode and fluent builder.

Mirrors ``src/protocol/metadata.rs`` in the Rust reference implementation.

Wire format per entry: ``[key: u8] [value_len: u16 BE] [value: N bytes]``.
Entries preserve insertion order; re-setting an existing key overwrites it
in place rather than appending a duplicate.
"""

from dataclasses import dataclass
from typing import Dict, Iterator, List, Optional, Tuple

from . import constants as _c
from .packet import InvalidPayloadError, Packet
from .types import PacketType


@dataclass
class MetadataEntry:
    """A single key-value metadata entry."""

    key: int
    value: bytes


MetadataMap = Dict[int, bytes]


class Metadata:
    """An ordered collection of metadata entries."""

    def __init__(self, entries: Optional[List[MetadataEntry]] = None) -> None:
        self._entries: List[MetadataEntry] = list(entries or [])

    @classmethod
    def from_map(cls, mapping: MetadataMap) -> "Metadata":
        obj = cls()
        for key, value in mapping.items():
            obj.set(key, value)
        return obj

    def set(self, key: int, value: bytes) -> None:
        value = bytes(value)
        for entry in self._entries:
            if entry.key == key:
                entry.value = value
                return
        self._entries.append(MetadataEntry(key, value))

    def set_string(self, key: int, value: str) -> None:
        self.set(key, value.encode("utf-8"))

    def set_u32(self, key: int, value: int) -> None:
        self.set(key, int(value).to_bytes(4, "big"))

    def set_u16(self, key: int, value: int) -> None:
        self.set(key, int(value).to_bytes(2, "big"))

    def set_u64(self, key: int, value: int) -> None:
        self.set(key, int(value).to_bytes(8, "big"))

    def get(self, key: int) -> Optional[bytes]:
        for entry in self._entries:
            if entry.key == key:
                return entry.value
        return None

    def get_string(self, key: int) -> Optional[str]:
        value = self.get(key)
        if value is None:
            return None
        try:
            return value.decode("utf-8")
        except UnicodeDecodeError:
            return None

    def get_u32(self, key: int) -> Optional[int]:
        value = self.get(key)
        if value is None or len(value) < 4:
            return None
        return int.from_bytes(value[:4], "big")

    def get_u16(self, key: int) -> Optional[int]:
        value = self.get(key)
        if value is None or len(value) < 2:
            return None
        return int.from_bytes(value[:2], "big")

    def get_u64(self, key: int) -> Optional[int]:
        value = self.get(key)
        if value is None or len(value) < 8:
            return None
        return int.from_bytes(value[:8], "big")

    def remove(self, key: int) -> bool:
        before = len(self._entries)
        self._entries = [e for e in self._entries if e.key != key]
        return len(self._entries) < before

    def contains(self, key: int) -> bool:
        return any(e.key == key for e in self._entries)

    def entries(self) -> List[MetadataEntry]:
        return list(self._entries)

    def is_empty(self) -> bool:
        return not self._entries

    def len(self) -> int:
        return len(self._entries)

    def __len__(self) -> int:
        return len(self._entries)

    def iter(self) -> Iterator[Tuple[int, bytes]]:
        for entry in self._entries:
            yield entry.key, entry.value

    def __iter__(self) -> Iterator[Tuple[int, bytes]]:
        return self.iter()

    def encode(self) -> bytes:
        out = bytearray()
        for entry in self._entries:
            out.append(entry.key)
            out += len(entry.value).to_bytes(2, "big")
            out += entry.value
        return bytes(out)

    @classmethod
    def decode(cls, data: bytes) -> "Metadata":
        obj = cls()
        offset = 0
        size = len(data)
        while offset < size:
            if size - offset < _c.METADATA_ENTRY_HEADER_SIZE:
                raise InvalidPayloadError("metadata entry truncated")
            key = data[offset]
            offset += 1
            value_len = int.from_bytes(data[offset:offset + 2], "big")
            offset += 2
            if offset + value_len > size:
                raise InvalidPayloadError("metadata value truncated")
            value = data[offset:offset + value_len]
            offset += value_len
            obj._entries.append(MetadataEntry(key, value))
        return obj

    def to_packet(self, seq: int, ts: int) -> Packet:
        return Packet.new(PacketType.METADATA, seq, ts, self.encode())

    @classmethod
    def from_packet(cls, packet: Packet) -> "Metadata":
        if packet.header.packet_type != PacketType.METADATA:
            raise InvalidPayloadError("not a metadata packet")
        return cls.decode(packet.payload)


class MetadataBuilder:
    """Fluent builder for common stream metadata (Rust ``MetadataBuilder``)."""

    def __init__(self) -> None:
        self._meta = Metadata()

    def _set(self, key: int, value: bytes) -> "MetadataBuilder":
        self._meta.set(key, value)
        return self

    def ssrc(self, value: int) -> "MetadataBuilder":
        return self._set(_c.META_SSRC, int(value).to_bytes(4, "big"))

    def track_title(self, title: str) -> "MetadataBuilder":
        return self._set(_c.META_TRACK_TITLE, title.encode("utf-8"))

    def artist(self, artist: str) -> "MetadataBuilder":
        return self._set(_c.META_ARTIST, artist.encode("utf-8"))

    def album(self, album: str) -> "MetadataBuilder":
        return self._set(_c.META_ALBUM, album.encode("utf-8"))

    def genre(self, genre: str) -> "MetadataBuilder":
        return self._set(_c.META_GENRE, genre.encode("utf-8"))

    def sample_rate(self, rate: int) -> "MetadataBuilder":
        return self._set(_c.META_SAMPLE_RATE, int(rate).to_bytes(4, "big"))

    def channels(self, channels: int) -> "MetadataBuilder":
        return self._set(_c.META_CHANNELS, int(channels).to_bytes(2, "big"))

    def codec_info(self, info: str) -> "MetadataBuilder":
        return self._set(_c.META_CODEC_INFO, info.encode("utf-8"))

    def bitrate(self, bitrate: int) -> "MetadataBuilder":
        return self._set(_c.META_BITRATE, int(bitrate).to_bytes(4, "big"))

    def duration_ms(self, duration: int) -> "MetadataBuilder":
        return self._set(_c.META_DURATION_MS, int(duration).to_bytes(8, "big"))

    def stream_title(self, title: str) -> "MetadataBuilder":
        return self._set(_c.META_STREAM_TITLE, title.encode("utf-8"))

    def stream_url(self, url: str) -> "MetadataBuilder":
        return self._set(_c.META_STREAM_URL, url.encode("utf-8"))

    def custom(self, key: int, value: bytes) -> "MetadataBuilder":
        if key < _c.META_CUSTOM_BASE:
            raise ValueError("custom metadata key must be >= 0x80")
        return self._set(key, value)

    def build(self) -> Metadata:
        return self._meta